### DISCUSSION TITLE
Issue with jq, discord notifactions are empty

### DISCUSSION THREAD

**Discussion by tim-shortons** [auanasgheps/snapraid-aio-script#131] [Q&A] [answered]
As the subject states, discord notifications are empty.  I can see the script start via discord notification "SnapRAID Script Job started", but the details for the remainder of the job are empty brackets.  Looking at the log I see the following several times with the first occurrence after the SMART report.  My assumption is a jq option is being called that is not recognized.  

This is the 3.4 beta version of the script

jq invocation from script:
`  if [ "$DISCORD" -eq 1 ]; then
  DISCORD_SUBJECT=$(echo "$NOTIFY_OUTPUT" | jq -Rs | cut -c 2- | rev | cut -c 2- | rev)
    curl -fsS -m 5 --retry 3 -o /dev/null -X POST \
    -H 'Content-Type: application/json' \
    -d '{"content": "'"$DISCORD_SUBJECT"'"}' \
    "$DISCORD_WEBHOOK_URL"
  fi`

```
jq - commandline JSON processor [version 1.5-1-a5b5cbe]
Usage: jq [options] <jq filter> [file...]

	jq is a tool for processing JSON inputs, applying the
	given filter to its JSON text inputs and producing the
	filter's results as JSON on standard output.
	The simplest filter is ., which is the identity filter,
	copying jq's input to its output unmodified (except for
	formatting).
	For more advanced filters see the jq(1) manpage ("man jq")
	and/or https://stedolan.github.io/jq

	Some of the options include:
	 -c		compact instead of pretty-printed output;
	 -n		use `null` as the single input value;
	 -e		set the exit status code based on the output;
	 -s		read (slurp) all inputs into an array; apply filter to it;
	 -r		output raw strings, not JSON texts;
	 -R		read raw strings, not JSON texts;
	 -C		colorize JSON;
	 -M		monochrome (don't colorize JSON);
	 -S		sort keys of objects on output;
	 --tab	use tabs for indentation;
	 --arg a v	set variable $a to value <v>;
	 --argjson a v	set variable $a to JSON value <v>;
	 --slurpfile a f	set variable $a to an array of JSON texts read from <f>;
	See the manpage for more options.
```

Additionally, I see there is a curl error at the end of the script; "curl: (22) The requested URL returned error: 400 ". I have healthcheck.io and discord notifications enabled.  I know the discord webhook url is correct as I receive the script job started notification.  Same with healthcheck.io I am receiving the snapraid events in the event dashboard.

I am on Ubuntu 18.04 with snapraid 11.6

**Comment by auanasgheps** [comment]
There was a fix for Discord notifications in 3.3.x, the recent changes in the beta might have broken it again. Unfortunately handling Discord in this script is very, very hard. Also I don't use it by myself so it's not the first thing I test 😛

I'll have a look

**Reply by tim-shortons** [reply]
Thanks for the response.  I'll disable discord notifications in the meantime.

**Reply by tim-shortons** [reply]
Just noticed another odd behavior.  After disabling the discord notifications in the config script, DISCORD=0, and leaving the discord webhook url populated it will send two emails after the job completes.
```
All jobs ended. [Thu Oct 17 12:45:25 EDT 2024]
Email address is set. Sending email report to **notreal@gmail.com** [Thu Oct 17 12:45:25 EDT 2024]
----------------------------------------
## Total time elapsed for SnapRAID: 0hrs 2min 23sec
Email address is set. Sending email report to **notreal@gmail.com** [Thu Oct 17 12:45:27 EDT 2024]
```
After replacing the discord webhook with the original value from the config, DISCORD_WEBHOOK_URL="your-webhook-url", it sends a single email on completion.
```
All jobs ended. [Thu Oct 17 14:42:40 EDT 2024]
----------------------------------------
## Total time elapsed for SnapRAID: 0hrs 4min 59sec
Email address is set. Sending email report to **notreal@gmail.com** [Thu Oct 17 14:42:40 EDT 2024]
```

**Reply by auanasgheps** [reply]
I got a feature request to integrate [Apprise](https://github.com/caronc/apprise), which is a broker to dozens of notification services including Discord. 
I believe that using this service I could solve all the weirdness around Discord notifications.

The disadvantage is that Apprise is self-hosted, you need to install another package and configure it. 

Would you like this approach? It would take more time though as I never used Apprise but looks very promising.

**Reply by tim-shortons** [reply]
Please don't go out of your way on my account, so long as I can receive email notifications that's more than enough for me.  I am currently using mtompkins script and we looking to replace it with your version and was playing around with the various notification options for fun.  I'll let users that require more complex notifications weigh in.  Thank you for the hard work you've put into this updated version.

**Reply by auanasgheps** [reply]
Thanks for the feedback, I will evaluate Apprise to offer more options.

**Comment by auanasgheps** [comment, marked as answer]
in my test environment Discord notifications are working correctly.
I am using the latest beta.

jq installed on my system (Debian 12) is 1.6, yours is 1.5 so this might be the difference. Your main OS not really updated, so some packages might not be aligned.
![immagine](https://github.com/user-attachments/assets/0e19ba03-37d6-4608-887d-59d25df15cf1)

![immagine](https://github.com/user-attachments/assets/4926f23e-eccb-406b-a6dd-830302756523)
